src/mapping/nav2_map_server/src/map_reps/occupancy_grid.cpp

```diff
@@ -0,0 +1,324 @@
+      "The map does not contain a resolution tag or it is invalid.");
+    throw std::runtime_error("The map does not contain a resolution tag or it is invalid.");
+  }
+
+  try {
+    doc["negate"] >> negate;
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain a negate tag or it is invalid.");
+    throw std::runtime_error("The map does not contain a negate tag or it is invalid.");
+  }
+
+  try {
+    doc["occupied_thresh"] >> occ_th;
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain an occupied_thresh tag or it is invalid.");
+    throw std::runtime_error("The map does not contain an occupied_thresh tag or it is invalid.");
+  }
+
+  try {
+    doc["free_thresh"] >> free_th;
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain a free_thresh tag or it is invalid.");
+    throw std::runtime_error("The map does not contain a free_thresh tag or it is invalid.");
+  }
+
+  try {
+    std::string modeS = "";
+    doc["mode"] >> modeS;
+    if (modeS == "trinary") {
+      mode = TRINARY;
+    } else if (modeS == "scale") {
+      mode = SCALE;
+    } else if (modeS == "raw") {
+      mode = RAW;
+    } else {
+      RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+        "Invalid mode tag \"%s\".", modeS.c_str());
+      throw std::runtime_error("Invalid mode tag.");
+    }
+  } catch (YAML::Exception) {
+    RCLCPP_DEBUG(rclcpp::get_logger("map_server"),
+      "The map does not contain a mode tag or it is invalid.");
+    mode = TRINARY;
+  }
+
+  try {
+    doc["origin"][0] >> origin[0];
+    doc["origin"][1] >> origin[1];
+    doc["origin"][2] >> origin[2];
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain an origin tag or it is invalid.");
+    throw std::runtime_error("The map does not contain an origin tag or it is invalid.");
+  }
+
+  try {
+    doc["image"] >> map_name_;
+    // TODO(bpwilcox): make this path-handling more robust
+    if (map_name_.size() == 0) {
+      RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+        "The image tag cannot be an empty string.");
+      throw std::runtime_error("The image tag cannot be an empty string.");
+    }
+    if (map_name_[0] != '/') {
+      // dirname can modify what you pass it
+      char * fname_copy = strdup(file_name.c_str());
+      map_name_ = std::string(dirname(fname_copy)) + '/' + map_name_;
+      free(fname_copy);
+    }
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain an image tag or it is invalid.");
+    throw std::runtime_error("The map does not contain an image tag or it is invalid.");
+  }
+}
+
+void OccGridLoader::LoadMapFromFile(std::string map_name_)
+{
+  SDL_Surface * img;
+  const char * name = map_name_.c_str();
+  unsigned char * pixels;
```

Comment:
I'd highly recommend not forward declaring these variables.  This was in the old map server and it made things really hard to read.

Reply:
Will consider redesign in future pull request

Reply:
I think moving them to before they're used will be a minimally invasive change if you're willing to do it now, otherwise please put a TODO or make a ticket.

Reply:
I will file an issue. Since this code is a refactor/port, this pull request is keeping most of the original code intact. As someone already suggested a further redesign of the code, I'd rather lump all these changes into the next pull request since I may change things a good bit.

Reply:
Opened in #82